README
$Id: README 69 2007-08-12 18:17:00Z jeff $

Author:
	Jeff Buchbinder <jeff@freemedsoftware.org>

FreeMED Electronic Medical Record and Practice Management System
Copyright (C) 1999-2013 FreeMED Software Foundation

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

-----------------------------------------------------------------------------

"libfreemed" is the set of language bindings allowing FreeMED methods to be
executed through the data relay debuting in version 0.9.0.




perl/Changes
Revision history for FreeMED::Relay

0.2.1	2007-08-23 10:06 EDT
	Fixes for a showstopping bug.
	Perl dependencies handled properly.

0.2     2007-08-13 20:26 EDT
        First public release



php/test.php
#!/usr/bin/php
<?php
 // $Id: test.php 60 2007-08-06 17:03:27Z jeff $
 //
 // Authors:
 //      Jeff Buchbinder <jeff@freemedsoftware.org>
 //
 // FreeMED Electronic Medical Record and Practice Management System
 // Copyright (C) 1999-2013 FreeMED Software Foundation
 //
 // This program is free software; you can redistribute it and/or modify
 // it under the terms of the GNU General Public License as published by
 // the Free Software Foundation; either version 2 of the License, or
 // (at your option) any later version.
 //
 // This program is distributed in the hope that it will be useful,
 // but WITHOUT ANY WARRANTY; without even the implied warranty of
 // MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 // GNU General Public License for more details.
 //
 // You should have received a copy of the GNU General Public License
 // along with this program; if not, write to the Free Software
 // Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

include_once(dirname(__FILE__).'/class.FreemedRelay.php');
openlog( 'libfreemed', LOG_PID | LOG_PERROR, LOG_LOCAL0 );

if ( $argc != 4 ) {
	print "syntax: $argv[0] url username password\n";
	print " example: \n";
	print "  $argv[0] http://localhost/freemed user myPassword\n";
	exit(0);
}

$r = new FreemedRelay( $argv[1] );
$r->SetCredentials( $argv[2], $argv[3] );
if ( $r->Login() ) {
	print "Connected to data relay!\n";
} else {
	die( "Failed to connect\n" );
}

// Now, quick test of the data relay
print "org.freemedsoftware.api.SystemConfig.GetAll : \n";
print_r( $r->Call( 'org.freemedsoftware.api.SystemConfig.GetAll' ) );
print "-------------------------------------------------------------------------- \n\n";
print "org.freemedsoftware.api.Scheduler.GetDailyAppointments : \n";
print_r( $r->Call( 'org.freemedsoftware.api.Scheduler.GetDailyAppointments', array ( '2007-08-01' ) ) );
print "-------------------------------------------------------------------------- \n\n";

?>


perl/0.9.x-test.pl
#!/usr/bin/perl
#
# $Id: 0.9.x-test.pl 108 2008-08-07 17:35:00Z jeff $
#
# Authors:
#      Jeff Buchbinder <jeff@freemedsoftware.org>
#
# FreeMED Electronic Medical Record and Practice Management System
# Copyright (C) 1999-2013 FreeMED Software Foundation
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

use lib 'lib';

use FreeMED::Relay;
use Data::Dumper;
use Config::Tiny;

my $config = Config::Tiny->read("0.9.x-config");
my $f = FreeMED::Relay->new( debug => 1 );
$f->set_credentials(
	$config->{'installation'}->{'uri'},
	$config->{'installation'}->{'username'},
	$config->{'installation'}->{'password'}
);
$f->login();

print "org.freemedsoftware.api.Scheduler.GetDailyAppointments()\n";
	print Dumper($f->call('org.freemedsoftware.api.Scheduler.GetDailyAppointments', '2007-01-12'));
	print "-"x50 . "\n";

print "org.freemedsoftware.api.PatientInterface.picklist('doe')\n";
	print Dumper($f->call('org.freemedsoftware.api.PatientInterface.picklist', 'doe'));
	print "-"x50 . "\n";



perl/numeric-test.pl
#!/usr/bin/perl
#
# $Id: numeric-test.pl 108 2008-08-07 17:35:00Z jeff $
#
# Authors:
#      Jeff Buchbinder <jeff@freemedsoftware.org>
#
# FreeMED Electronic Medical Record and Practice Management System
# Copyright (C) 1999-2008 FreeMED Software Foundation
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

use lib 'lib';

use FreeMED::Relay;
use Data::Dumper;
use Config::Tiny;

my $config = Config::Tiny->read("0.9.x-config");
my $f = FreeMED::Relay->new( debug => 1 );
$f->set_credentials(
	$config->{'installation'}->{'uri'},
	$config->{'installation'}->{'username'},
	$config->{'installation'}->{'password'}
);
$f->login();

print "org.freemedsoftware.api.PatientInterface.NumericSearch( ... )\n";
	print Dumper($f->call('org.freemedsoftware.api.PatientInterface.NumericSearch', { last_name => '272', first_name => '86', year_of_birth => '1937' } ));
	print "-"x50 . "\n";



csharp/Relay.cs
/*
 * $Id: Relay.cs 44 2007-02-06 13:14:22Z jeff $
 *
 * Authors:
 * 	Jeff Buchbinder <jeff@freemedsoftware.org>
 *
 * FreeMED Electronic Medical Record and Practice Management System
 * Copyright (C) 1999-2013 FreeMED Software Foundation
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

using System;
using System.IO;
using System.Net;
using System.Reflection;
using System.Text;
using System.Web;

using Newtonsoft.Json;

namespace org.freemedsoftware.FreeMED {

public class Relay {
	private string _uri;
	private string _username;
	private string _password;
	private bool _debug;
	private CookieCollection _cookies;

	public Relay ( ) {
		_cookies = null;
		_uri = null;
		_username = null;
		_password = null;
		_debug = false;
	}

	public bool Debug {
		get { return _debug; }
		set { _debug = value; }
	}
	
	public void SetParameters ( string s_uri, string s_username, string s_password ) {
		_uri = s_uri;
		_username = s_username;
		_password = s_password;
	} // end method SetParameters

	public bool Login ( ) {
		if ( (_uri.Length < 5) || (_username.Length < 3) ) {
			Console.WriteLine("Need to call SetParameters()");
			return false;
		}
		return (bool) Call( 
			"org.freemedsoftware.public.Login.Validate",
			new object[] {
				_username,
				_password
			}
		);
	} // end method Login

	public object Call ( string method, object[] parameters ) {
		return Call ( method, parameters, null );
	}
	public object Call ( string method, object[] parameters, Type type ) {
		string outputValue = CallRaw( method, parameters, type );
		object deserializedObject;
		if ( type == null ) {
			deserializedObject = (object) JavaScriptConvert.DeserializeObject(outputValue);
		} else {
			deserializedObject = JavaScriptConvert.DeserializeObject(outputValue, type);
		}
		if ( _debug ) { Console.WriteLine( "DEBUG: deserialized = {0}", deserializedObject.ToString() ); }
		return deserializedObject;
	}
	public string CallRaw ( string method, object[] parameters, Type type ) {
		string fullUri = String.Format( "{0}/relay.php/json/{1}", _uri, method );
		if ( _debug ) { Console.WriteLine( "DEBUG: POST to {0} with {1}", fullUri, parameters.ToString() ); }
		HttpWebRequest webRequest = (HttpWebRequest) HttpWebRequest.Create( fullUri );
		webRequest.ContentType = "application/x-www-form-urlencoded";
		webRequest.Method = "POST";
		webRequest.CookieContainer = new CookieContainer();
		if (_cookies != null) {
			webRequest.CookieContainer.Add( _cookies );
		}

		// Form parameters
		string formedParameters = "";
		for (int i=0; i<parameters.Length; i++) {
			Type objectType = parameters[i].GetType();
			if ( _debug ) { Console.WriteLine("DEBUG: Processing parameter {0} type = {1}", i, objectType.ToString() ); }
			if ( formedParameters.Length > 0 ) { formedParameters += "&"; }
			switch (objectType.ToString()) {
				case "System.Int32":
				formedParameters += "param" + i.ToString() + "=" + parameters[i].ToString();
				break;

				case "System.String":
				formedParameters += "param" + i.ToString() + "=" + parameters[i];
				break;

				default:
				if ( _debug ) { Console.WriteLine( "DEBUG: serializing non-understood parameter" ); }
				// Check for ToJson() method
				try {
					MethodInfo[] mi = objectType.GetMethods();
					bool found = false;
					foreach (MethodInfo m in mi) {
						if ( m.Name == "ToJson" ) {
							if ( _debug ) { Console.WriteLine( "DEBUG: found ToJson() method" ); }
							found = true;
						}
					}
					if (found) {
						MethodInfo ToJson = objectType.GetMethod ("ToJson");
						formedParameters += "param" + i.ToString() + "=" + System.Web.HttpUtility.UrlEncode( (string) ToJson.Invoke (parameters[i], null) );
					} else {
						throw new Exception( "parameter" );
					}
				} catch (Exception) {
					formedParameters += "param" + i.ToString() + "=" + (string) JavaScriptConvert.SerializeObject( parameters[i] );
				}
				break;
			}
		}

		if ( _debug ) { Console.WriteLine( "DEBUG: formedParameters = '{0}'", formedParameters ); }
		byte[] bytes = Encoding.ASCII.GetBytes( formedParameters );
		Stream os = null;
		try {
			webRequest.ContentLength = bytes.Length;
			os = webRequest.GetRequestStream( );
			os.Write( bytes, 0, bytes.Length );
		} catch ( WebException ex ) {
			Console.WriteLine( "ERROR: Call() caught exception for {0} ({1})", fullUri, ex.Message );
			return null;
		} finally {
			if (os != null) { os.Close( ); }
		}

		try {
			HttpWebResponse webResponse = (HttpWebResponse) webRequest.GetResponse( );
			if ( webResponse == null ) { return null; }
			StreamReader sr = new StreamReader( webResponse.GetResponseStream( ) );
			string outputValue = sr.ReadToEnd( ).Trim( );
			if ( _debug ) { Console.WriteLine( "DEBUG: Returned {0}", outputValue ); }

			// Pass cookie value back
			if (webResponse.Cookies.Count > 0) {
				_cookies = webResponse.Cookies;
			}
			if ( _debug ) { Console.WriteLine( "DEBUG: Finished importing new cookies, if any" ); }

			try {
				if ( _debug ) { Console.WriteLine( "DEBUG: About to examine outputValue" ); }
				if ( ( outputValue == null ) || ( outputValue == "" ) ) {
					if ( _debug ) { Console.WriteLine( "DEBUG: NULL output value found" ); }
					return "false";
				}
				if ( _debug ) { Console.WriteLine( "DEBUG: About to return outputValue" ); }
				return outputValue;
			} catch (NullReferenceException iEx) {
				Console.WriteLine( "ERROR: Caught exception for {0}", iEx.Message );
				return "";
			}
		} catch ( WebException ex ) {
			Console.WriteLine( "ERROR: Caught exception for {0} ({1})", fullUri, ex.Message );
		} 

		return null;
	} // end Call

} // end class

} // end namespace



perl/Makefile.PL
#!/usr/bin/perl
#
# $Id: Makefile.PL 108 2008-08-07 17:35:00Z jeff $
#
# Authors:
#      Jeff Buchbinder <jeff@freemedsoftware.org>
#
# FreeMED Electronic Medical Record and Practice Management System
# Copyright (C) 1999-2008 FreeMED Software Foundation
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME                => 'FreeMED::Relay',
    AUTHOR              => 'Jeff Buchbinder <RTFIREFLY@cpan.org>',
    VERSION_FROM        => 'lib/FreeMED/Relay.pm',
    ABSTRACT_FROM       => 'lib/FreeMED/Relay.pm',
    PL_FILES            => {},
    PREREQ_PM => {
	'JSON' => '1.14',
	'LWP::UserAgent' => '2.033',
        'Test::More' => 0,
    },
    dist                => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean               => { FILES => 'FreeMED-Relay-*' },
);



csharp/RelayTest.cs
/*
 * $Id: RelayTest.cs 88 2007-09-06 01:44:08Z jeff $
 *
 * Authors:
 * 	Jeff Buchbinder <jeff@freemedsoftware.org>
 *
 * FreeMED Electronic Medical Record and Practice Management System
 * Copyright (C) 1999-2013 FreeMED Software Foundation
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

using System;
using System.Collections;
using System.IO;
using System.Net;
using System.Reflection;
using System.Text;
using System.Web;

using Newtonsoft.Json;
using org.freemedsoftware.FreeMED;

namespace org.freemedsoftware.FreeMED {

class RelayTest {
	public class ApptRecord {
		public ApptRecord ( ) { }

		public string date_of;
		public string date_of_mdy;
		public string hour;
		public string minute;
		public string appointment_time;
		public string provider;
		public int provider_id;
		public string patient;
		public int patient_id;
		public string note;
		public string status;
		public string status_color;
		public int scheduler_id;
		public int duration;
		public string resource_type;
	}

	public static int Main ( string[] argv ) {
		Console.WriteLine("org.freemedsoftware.FreeMED.Relay Test Suite $Revision: 88 $");
		if ( argv.Length != 3 ) {
			Console.WriteLine("syntax : program URL username password");
			return 1;
		}
		Console.WriteLine(" * Attempting login");
		Console.WriteLine(" - Creating Relay instance");
		Relay r = new Relay();
		Console.WriteLine(" - Debug = true");
		r.Debug = true;
		Console.WriteLine(" - SetParameters ( {0}, {1}, {2} )", argv[0], argv[1], argv[2]);
		r.SetParameters( argv[0], argv[1], argv[2] );
		Console.WriteLine(" - Login( )");
		r.Login();
		Console.WriteLine(" - Call org.freemedsoftware.api.Scheduler.GetDailyAppointments");
		string val = r.CallRaw( "org.freemedsoftware.api.Scheduler.GetDailyAppointments", new object[] { "2006-01-12" }, null );
		ApptRecord[] x = (ApptRecord []) JavaScriptConvert.DeserializeObject( val, typeof(ApptRecord[]) );
		Console.WriteLine( "ApptResults.Length = " + x.Length );
		for (int i=0; i<x.Length; i++) {
			Console.WriteLine(" - Patient : {0} ({1}), Date : {2}", x[i].patient, x[i].patient_id, x[i].date_of );
		}
		return 0;
	}
}

} // end namespace



csharp/AssemblyInfo.cs
/*
 * $Id: AssemblyInfo.cs 39 2007-01-26 00:39:48Z jeff $
 *
 * Authors:
 * 	Jeff Buchbinder <jeff@freemedsoftware.org>
 *
 * FreeMED Electronic Medical Record and Practice Management System
 * Copyright (C) 1999-2013 FreeMED Software Foundation
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

using System.Reflection;
using System;
using System.Runtime.CompilerServices;

// Information about this assembly is defined by the following
// attributes.
//
// change them to the information which is associated with the assembly
// you compile.

[assembly: AssemblyTitle("FreeMED.NET")]
[assembly: AssemblyDescription("FreeMED Relay .NET bindings")]
[assembly: AssemblyConfiguration("")]
[assembly: AssemblyCompany("FreeMED Software Foundation")]
[assembly: AssemblyProduct("")]
[assembly: AssemblyCopyright("")]
[assembly: AssemblyTrademark("")]
[assembly: AssemblyCulture("")]

// The assembly version has following format :
//
// Major.Minor.Build.Revision
//
// You can specify all values by your own or you can build default build and revision
// numbers with the '*' character (the default):

[assembly: AssemblyVersion("0.0.1")]

// The following attributes specify the key for the sign of your assembly. See the
// .NET Framework documentation for more information about signing.
// This is not required, if you don't want signing let these attributes like they're.
[assembly: AssemblyDelaySign(false)]
[assembly: AssemblyKeyFile("")]



python/FreemedRelay.py
#!/usr/bin/python
#
# $Id: FreemedRelay.py 87 2007-09-06 01:19:23Z jeff $
#
# Authors:
#      Jeff Buchbinder <jeff@freemedsoftware.org>
#
# FreeMED Electronic Medical Record and Practice Management System
# Copyright (C) 1999-2013 FreeMED Software Foundation
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

import os.path
import sys
import simplejson
import urllib, MultipartPostHandler
from UserDict import UserDict

class FreemedRelay(UserDict):
	"FreeMED 0.9.x+ data relay binding"
	def __init__( self, url, debug=False ):
		UserDict.__init__( self )
		self['url'] = url

		self['COOKIEFILE'] = 'freemed-cookies.lwp'

		if debug:
			import logging
			logger = logging.getLogger("cookielib")
			logger.addHandler(logging.StreamHandler(sys.stdout))
			logger.setLevel(logging.DEBUG)
	
		cj = None
		ClientCookie = None
		cookielib = None
	
		try:
			import cookielib            
		except ImportError:
			pass
		else:
			import urllib2    
			urlopen = urllib2.urlopen
			cj = cookielib.LWPCookieJar()
			Request = urllib2.Request

		if not cookielib:
			try:                                            
				import ClientCookie 
			except ImportError:
				import urllib2
				urlopen = urllib2.urlopen
				Request = urllib2.Request
			else:
				urlopen = ClientCookie.urlopen
				cj = ClientCookie.LWPCookieJar()
				Request = ClientCookie.Request
		if cj != None:
			if os.path.isfile( self['COOKIEFILE'] ):
				if debug:
					print 'DEBUG: Loading cookiefile ' + self['COOKIEFILE']
				cj.load( self['COOKIEFILE'] )
			if cookielib:
				opener = urllib2.build_opener(urllib2.HTTPCookieProcessor(cj), MultipartPostHandler.MultipartPostHandler)
				urllib2.install_opener(opener)
			else:
				opener = ClientCookie.build_opener(ClientCookie.HTTPCookieProcessor(cj), MultipartPostHandler.MultipartPostHandler)
				ClientCookie.install_opener(opener)
		self['Request'] = Request
		self['urlopen'] = urlopen
		self['cj'] = cj

	def LoadURL( self, method, data=None, postFiles=None, debug=False ):
		if data is None:
			txdata = None
		else:
			txdata = { }
			count = 0
			for iter in data:
				txdata[ 'param' + str( count ) ] = simplejson.dumps( iter )
				count += 1
			txdata = urllib.urlencode( txdata )
			if debug:
				print 'DEBUG: ' + txdata

		if postFiles is not None:
			for (k, v) in postFiles:
				txdata[ k ] = open( v )
				if debug:
					print 'DEBUG (POST): ' + k + ' = ' + v
	
		txheaders =  {'User-agent' : 'Mozilla/4.0 (compatible; MSIE 5.5; Windows NT)'}
	
		try:
			fullurl = self['url'] + method
			if debug:
				print 'DEBUG: fullurl = ' + fullurl
			req = self['Request']( fullurl, txdata, txheaders )
			handle = self['urlopen']( req )
			val = handle.read()
		except IOError, e:
			print 'We failed to open "%s".' % fullurl
			if hasattr(e, 'code'):
				print 'We failed with error code - %s.' % e.code
			elif hasattr(e, 'reason'):
				print "The error object has the following 'reason' attribute :", e.reason
				print "This usually means the server doesn't exist, is down, or we don't have an internet connection."
			else:
				print 'Here are the headers of the page :'
				print handle.info()
		
		print ' '
		if self['cj'] == None:
			print "We don't have a cookie library available - sorry."
			print "I can't show you any cookies."
			sys.exit()
		else:
			if debug:
				print 'These are the cookies we have received so far :'
				for index, cookie in enumerate(self['cj']):
					print index, '  :  ', cookie
				print 'DEBUG: save cookies in ' + self['COOKIEFILE']
			self['cj'].save( self['COOKIEFILE'] )
			try:
				ret = simplejson.loads( val )
			except ValueError:
				print 'ERROR, RETURNED: ' + val
				ret = False
			return ret

#
#	Test suite goes here:::
#

if __name__ == "__main__":
	BASE = 'http://localhost/freemed'
	RELAY_PATH = '/relay.php/json/'
	
	x = FreemedRelay( BASE + RELAY_PATH )
	print 'org.freemedsoftware.public.Login.Validate: '
	print x.LoadURL( 'org.freemedsoftware.public.Login.Validate', [ 'demo', 'demo' ] )
	print "\n----------------------------------------\n"
	print 'org.freemedsoftware.api.Scheduler.GetDailyAppointments("2007-01-12"): '
	print x.LoadURL( 'org.freemedsoftware.api.Scheduler.GetDailyAppointments', ['2007-01-02'] )
	print "\n----------------------------------------\n"
	print 'org.freemedsoftware.api.Messages.view(): '
	print x.LoadURL( 'org.freemedsoftware.api.Messages.view' )
	print "\n----------------------------------------\n"



php/class.FreemedRelay.php
<?php
 // $Id: class.FreemedRelay.php 60 2007-08-06 17:03:27Z jeff $
 //
 // Authors:
 //      Jeff Buchbinder <jeff@freemedsoftware.org>
 //
 // FreeMED Electronic Medical Record and Practice Management System
 // Copyright (C) 1999-2013 FreeMED Software Foundation
 //
 // This program is free software; you can redistribute it and/or modify
 // it under the terms of the GNU General Public License as published by
 // the Free Software Foundation; either version 2 of the License, or
 // (at your option) any later version.
 //
 // This program is distributed in the hope that it will be useful,
 // but WITHOUT ANY WARRANTY; without even the implied warranty of
 // MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 // GNU General Public License for more details.
 //
 // You should have received a copy of the GNU General Public License
 // along with this program; if not, write to the Free Software
 // Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

class FreemedRelay {

	protected $uri;
	protected $username;
	protected $password;
	protected $cookiefile;

	public function __construct ( $uri ) {
		$this->uri = $uri . "/relay.php/json/";
		$this->cookiefile = tempnam( '/tmp', 'freemedrelay' );
		$this->loggedin = false;
	} // end constructor

	public function SetCredentials ( $username, $password ) {
		$this->username = $username;
		$this->password = $password;
	} // end method SetCredentials

	// Method: Login
	//
	//	Login to FreeMED 0.9.x+ data relay.
	//
	public function Login ( ) {
		$res = $this->curl_call( $this->uri.'org.freemedsoftware.public.Login.Validate', array ( $this->username, $this->password ) );
		$res = json_decode( $res );
		if ( $res ) {
			$this->loggedin = true;
		}
		return $res;
	} // end method Login

	// Method: Call
	//
	//	Call a remote method through the FreeMED 0.9.x+ data relay.
	//
	// Parameters:
	//
	//	$namespace - Relay namespace. (example: 'org.freemedsoftware.public.Login.Validate')
	//
	//	$params - (optional) Array of parameters
	//
	//	$files - (optional) Hash of files for multipart attachments
	//
	// Returns:
	//
	//	PHP object or other encoding.
	//
	public function Call ( $namespace, $params = NULL, $files = NULL ) {
		$url = $this->uri . $namespace;
		$res = json_decode( $this->curl_call( $url, $params, $files ) );
		return $res;
	} // end method Call

	// Method: curl_call
	//
	//	Low-level cURL call.
	//
	// Parameters:
	//
	//	$url - Fully qualified URL
	//
	//	$params - (optional) Array of parameters
	//
	//	$files - (optional) Hash of file variables for multipart attachment
	//
	// Returns:
	//
	//	JSON-encoded body.
	//
	protected function curl_call ( $url, $params = NULL, $files = NULL ) {
		//print "Using $url\n";
		if ( ( strpos( $url, '.public.') === false ) and ! $this->loggedin ) {
			syslog( LOG_ERR, "Need to be logged in first" );
			return false;
		}

		$c = curl_init( );
		curl_setopt( $c, CURL_TIMEOUT, 120 );
		curl_setopt( $c, CURLOPT_URL, $url ); 
		curl_setopt( $c, CURLOPT_HEADER, 0 ); 
		curl_setopt( $c, CURLOPT_RETURNTRANSFER, 1 ); 
		curl_setopt( $c, CURLOPT_POST, 1 );
		$p = array ( );
		if ( $params != NULL ) {
			$p = array_merge( $p, $this->encode_parameters( $params ) ); 
		}
		if ( $files != NULL ) {
			foreach ( $files AS $v => $f ) {
				$m[ $v ] = '@' . $f; 
			}
			$p = array_merge( $p, $m );
		}
		curl_setopt( $c, CURLOPT_POSTFIELDS, $p ); 
		curl_setopt( $c, CURLOPT_COOKIEFILE, $this->cookiefile ); 
		curl_setopt( $c, CURLOPT_COOKIEJAR, $this->cookiefile ); 
		$data = curl_exec( $c );
		curl_close( $c );
		//print "[ DEBUG : $data ]\n";
		return $data;
	} // end method curl_call

	// Method: encode_parameters
	//
	//	Prepare parameters for the data relay.
	//
	// Parameters:
	//
	//	$params - Array of parameters to be passed.
	//
	// Returns:
	//
	//	Hash of parameters for POST-ing to the data relay.
	//
	protected function encode_parameters ( $params ) {
		// Form params
		if (is_array( $params )) {
			for ( $i=0; $i<count($params); $i++ ) {
				$f['param'.( $i + 0 )] = json_encode( $params[$i] );
			}
		} else {
			$f = '';
		}
		//print_r($f);
		return $f;
	} // end method encode_parameters

} // end class FreemedRelay

?>


perl/lib/FreeMED/Relay.pm
#!/usr/bin/perl
#
# $Id: Relay.pm 107 2008-08-07 15:13:22Z jeff $
#
# Authors:
#      Jeff Buchbinder <jeff@freemedsoftware.org>
#
# FreeMED Electronic Medical Record and Practice Management System
# Copyright (C) 1999-2006 FreeMED Software Foundation
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

#
#	FreeMED::Relay package for communicating with FreeMED 0.9.x+
#	

package FreeMED::Relay;

use LWP::UserAgent;
use HTTP::Request::Common;
use JSON;	# objToJson(), jsonToObj()
use HTTP::Cookies;

use vars qw{ $VERSION };
BEGIN {
	$VERSION = '0.2.3';
}

sub new {
	my $class = shift;
	my $debug = shift;
	my $self = {};
	bless $self, $class;
	$self->{'debug'} = $debug;
	$self->_init();
	$JSON::UnMapping = 1;
	$JSON::QuotApos = 1;
	$JSON::BareKey = 1;
	return $self;
}

sub set_credentials {
	my $self = shift;

	my ( $base_uri, $username, $password ) = @_;
	print "base_uri = $base_uri\n" if $self->{'debug'};
	print "username = $username\n" if $self->{'debug'};
	print "password = $password\n" if $self->{'debug'};
	$self->{base_uri} = $base_uri;
	$self->{username} = $username;
	$self->{password} = $password;
}

sub login {
	my $self = shift;
	if (!defined($self->{username}) or !defined($self->{password})) {
		die "login(): Must set credentials before logging in\n";
	}

	$self->_init() if (!defined($self->{ua}));
	my $login = $self->call(
		'org.freemedsoftware.public.Login.Validate',
		$self->{'username'},
		$self->{'password'},
	);
	$self->{'logged_in'} = true;
	return $login;
}

sub call {
	my $self = shift;
	my $method = shift;
	my @params = @_;

	if (!($method =~ /public/) && !$self->{'logged_in'}) {
		print "Must be logged in first\n" if $self->{'debug'};
		return undef;
	}

	my $count = 0; my %p;
	foreach my $param (@params) {
		print "param = '$param'\n";
		if ( $param =~ /^HASH\(/ && $param->{'@var'} ) {
			print "Found file upload var in $param->{'@var'}\n" if ($self->{'debug'});
			# Add file transfer under @var = var, @filename = filename
			$p{$param->{'@var'}} = [ $param->{'@filename'} ];
		} else {
			my $j = new JSON;
			my $json = $j->allow_nonref->encode( $param );
			print "param = $param, count = $count, json = $json\n" if $self->{'debug'};
			$p{"param${count}"} = ($json ? $json : $param );
			$count++;
		}
	}
	my $res = $self->{ua}->request(
		POST $self->{base_uri}."/relay.php/json/${method}",
		Content_Type => 'form-data',
		Content => [ %p ]
	);
	$self->{'cookie_jar'}->save();
	print "content : ".$res->content."\n" if ($self->{debug});
	my $j = new JSON;
	return $j->allow_nonref->decode( $res->content );
}

sub _init {
	my $self = shift;
	$self->{'ua'} = LWP::UserAgent->new;
	$self->{'cookie_jar'} = HTTP::Cookies->new( 'autosave' => 1 );
	$self->{'ua'}->cookie_jar( $self->{'cookie_jar'} );
}

1;
__END__

=pod

=head1 NAME

FreeMED::Relay

=head1 SYNOPSIS

Provide access to FreeMED 0.9.x+ data relay

=head1 DESCRIPTION

Backend access to the FreeMED Electronic Medical Record and Practice
Management system ( http://freemedsoftware.org/ ) for versions begining
with 0.9.0 using the JSON transport capabilities of its data relay.

=head1 METHODS

=over 4

=item new ( %options )

Returns a FreeMED::Relay object.

C<new> takes "debug" as a boolean argument.

=item set_credentials ( $base_uri, $username, $password )

Sets the credentials used to access the FreeMED installation in question. The
C<$base_uri> variable should be the base name of the installation, such as
"http://localhost/freemed".

=item login ( )

Log into the specified installation of FreeMED. Returns true or false
depending on whether or not it is successful.

=item call ( $method, $params ... )

Execute a remote procedural call, translating to and from JSON transparently.
If an argument is a hash with the keys '@var' and '@filename' it is assumed
that the filename in question will be uploaded and attached to the form
variable '@var'.

=item _init ( )

Internal method for initializing the LWP user agent, cookie jar and other
special things.

=cut

